include/Sema/TyMapperVisitor.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef GLU_SEMA_TY_MAPPER_VISITOR_HPP
+#define GLU_SEMA_TY_MAPPER_VISITOR_HPP
+
+#include "Basic/Diagnostic.hpp"
+#include "ConstraintSystem.hpp"
+#include "TypeMapper.hpp"
+
+namespace glu::sema {
+
+/// A reusable type mapping base for substituting/rewriting types using a
+/// solution.
+template <typename Derived>
+class TypeMappingVisitorBase
+    : public glu::sema::TypeMapper<Derived>,
+      public glu::types::TypeVisitor<Derived, glu::types::TypeBase *> {
+protected:
+    glu::DiagnosticManager &_diagManager;
+    glu::ast::ASTContext *_context;
+
+public:
+    TypeMappingVisitorBase(
+        glu::DiagnosticManager &diagManager, glu::ast::ASTContext *context
+    )
+        : _diagManager(diagManager), _context(context)
+    {
+    }
+
+    glu::types::TypeBase *visitTypeBase(glu::types::TypeBase *type)
+    {
+        return type;
+    }
+
+    glu::types::TypeBase *visitFunctionTy(glu::types::FunctionTy *type)
+    {
+        auto *returnType
+            = static_cast<Derived *>(this)->visit(type->getReturnType());
```

Comment:
you don't need any of those `static_cast`, visit is going to do the asImpl itself
```suggestion
        auto *returnType = visit(type->getReturnType());
```